File: packages/callwave_flutter_method_channel/ios/Classes/CallwaveMethodHandler.swift
import Foundation
import Flutter

final class CallwaveMethodHandler {
  private let callManager: IOSCallManager

  init(callManager: IOSCallManager) {
    self.callManager = callManager
  }

  func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      result(nil)

    case "showIncomingCall":
      guard
        let args = call.arguments as? [String: Any],
        let payload = CallPayload(dictionary: args)
      else {
        result(FlutterError(code: "invalid_payload", message: "Missing incoming payload", details: nil))
        return
      }
      callManager.showIncomingCall(payload)
      result(nil)

    case "showOutgoingCall":
      guard
        let args = call.arguments as? [String: Any],
        let payload = CallPayload(dictionary: args)
      else {
        result(FlutterError(code: "invalid_payload", message: "Missing outgoing payload", details: nil))
        return
      }
      callManager.showOutgoingCall(payload)
      result(nil)

    case "registerBackgroundIncomingCallValidator":
      let args = call.arguments as? [String: Any]
      guard
        let dispatcherHandle = (args?["backgroundDispatcherHandle"] as? NSNumber)?.int64Value
      else {
        result(
          FlutterError(
            code: "invalid_background_validator",
            message: "Dispatcher handle is required",
            details: nil
          )
        )
        return
      }
      let acceptCallbackHandle = (args?["backgroundCallbackHandle"] as? NSNumber)?.int64Value
      let declineCallbackHandle =
        (args?["backgroundDeclineCallbackHandle"] as? NSNumber)?.int64Value
      guard acceptCallbackHandle != nil || declineCallbackHandle != nil else {
        result(
          FlutterError(
            code: "invalid_background_validator",
            message: "At least one callback handle is required",
            details: nil
          )
        )
        return
      }
      callManager.registerBackgroundIncomingCallValidator(
        backgroundDispatcherHandle: dispatcherHandle,
        backgroundAcceptCallbackHandle: acceptCallbackHandle,
        backgroundDeclineCallbackHandle: declineCallbackHandle
      )
      result(nil)

    case "clearBackgroundIncomingCallValidator":
      callManager.clearBackgroundIncomingCallValidator()
      result(nil)

    case "endCall":
      let args = call.arguments as? [String: Any]
      guard let callId = args?["callId"] as? String else {
        result(FlutterError(code: "invalid_call_id", message: "callId is required", details: nil))
        return
      }
      callManager.endCall(callId: callId)
      result(nil)

    case "acceptCall":
      let args = call.arguments as? [String: Any]
      guard let callId = args?["callId"] as? String else {
        result(FlutterError(code: "invalid_call_id", message: "callId is required", details: nil))
        return
      }
      guard callManager.acceptCall(callId: callId) else {
        result(
          FlutterError(
            code: "invalid_call_id",
            message: "No active incoming call found for callId=\(callId)",
            details: nil
          )
        )
        return
      }
      result(nil)

    case "confirmAcceptedCall":
      let args = call.arguments as? [String: Any]
      guard let callId = args?["callId"] as? String else {
        result(FlutterError(code: "invalid_call_id", message: "callId is required", details: nil))
        return
      }
      guard callManager.confirmAcceptedCall(callId: callId) else {
        result(
          FlutterError(
            code: "invalid_call_id",
            message: "No active accepted call found for callId=\(callId)",
            details: nil
          )
        )
        return
      }
      result(nil)

    case "declineCall":
      let args = call.arguments as? [String: Any]
      guard let callId = args?["callId"] as? String else {
        result(FlutterError(code: "invalid_call_id", message: "callId is required", details: nil))
        return
      }
      guard callManager.declineCall(callId: callId) else {
        result(
          FlutterError(
            code: "invalid_call_id",
            message: "No active incoming call found for callId=\(callId)",
            details: nil
          )
        )
        return
      }
      result(nil)

    case "markMissed":
      let args = call.arguments as? [String: Any]
      guard let callId = args?["callId"] as? String else {
        result(FlutterError(code: "invalid_call_id", message: "callId is required", details: nil))
        return
      }
      let extra = args?["extra"] as? [String: Any]
      callManager.markMissed(callId: callId, extra: extra)
      result(nil)

    case "syncCallConnectedState":
      let args = call.arguments as? [String: Any]
      guard
        let callId = args?["callId"] as? String,
        let connectedAtMs = (args?["connectedAtMs"] as? NSNumber)?.int64Value,
        connectedAtMs > 0
      else {
        result(
          FlutterError(
            code: "invalid_connected_state",
            message: "callId and connectedAtMs are required",
            details: nil
          )
        )
        return
      }
      callManager.syncCallConnectedState(callId: callId, connectedAtMs: connectedAtMs)
      result(nil)

    case "clearCallState":
      let args = call.arguments as? [String: Any]
      guard let callId = args?["callId"] as? String else {
        result(FlutterError(code: "invalid_call_id", message: "callId is required", details: nil))
        return
      }
      callManager.clearCallState(callId: callId)
      result(nil)

    case "getActiveCallIds":
      result(callManager.activeCallIds())

    case "getActiveCallEventSnapshots":
      result(callManager.activeCallEventSnapshots())

    case "syncActiveCallsToEvents":
      callManager.syncActiveCallsToEvents()
      result(nil)

    case "requestNotificationPermission":
      callManager.requestNotificationPermission(result: result)

    case "takePendingStartupAction":
      result(callManager.takePendingStartupAction())

    case "requestFullScreenIntentPermission":
      result(nil)

    case "setPostCallBehavior":
      let args = call.arguments as? [String: Any]
      let behavior = args?["postCallBehavior"] as? String
      callManager.setPostCallBehavior(rawValue: behavior)
      result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

File: packages/callwave_flutter_method_channel/android/src/main/kotlin/com/callwave/flutter/methodchannel/CallwaveMethodHandler.kt
package com.callwave.flutter.methodchannel

import android.app.Activity
import com.callwave.flutter.methodchannel.manager.AndroidCallManager
import com.callwave.flutter.methodchannel.model.CallPayload
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class CallwaveMethodHandler(
    private val callManager: AndroidCallManager,
) : MethodChannel.MethodCallHandler {
    var activity: Activity? = null

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> {
                callManager.initialize()
                result.success(null)
            }

            "showIncomingCall" -> {
                val payload = payloadFromCall(call)
                if (payload == null) {
                    result.error("invalid_payload", "Missing incoming call payload", null)
                    return
                }
                callManager.showIncomingCall(payload)
                result.success(null)
            }

            "showOutgoingCall" -> {
                val payload = payloadFromCall(call)
                if (payload == null) {
                    result.error("invalid_payload", "Missing outgoing call payload", null)
                    return
                }
                callManager.showOutgoingCall(payload)
                result.success(null)
            }

            "registerBackgroundIncomingCallValidator" -> {
                val dispatcherHandle = call.argument<Number>(
                    CallwaveConstants.EXTRA_BACKGROUND_DISPATCHER_HANDLE,
                )?.toLong()
                val callbackHandle = call.argument<Number>(
                    CallwaveConstants.EXTRA_BACKGROUND_CALLBACK_HANDLE,
                )?.toLong()
                if (dispatcherHandle == null || callbackHandle == null) {
                    result.error(
                        "invalid_background_validator",
                        "Dispatcher and callback handles are required",
                        null,
                    )
                    return
                }
                callManager.registerBackgroundIncomingCallValidator(
                    backgroundDispatcherHandle = dispatcherHandle,
                    backgroundCallbackHandle = callbackHandle,
                )
                result.success(null)
            }

            "clearBackgroundIncomingCallValidator" -> {
                callManager.clearBackgroundIncomingCallValidator()
                result.success(null)
            }

            "endCall" -> {
                val callId = call.argument<String>(CallwaveConstants.EXTRA_CALL_ID)
                if (callId.isNullOrBlank()) {
                    result.error("invalid_call_id", "callId is required", null)
                    return
                }
                callManager.endCall(callId)
                result.success(null)
            }

            "acceptCall" -> {
                val callId = call.argument<String>(CallwaveConstants.EXTRA_CALL_ID)
                if (callId.isNullOrBlank()) {
                    result.error("invalid_call_id", "callId is required", null)
                    return
                }
                if (!callManager.acceptCall(callId)) {
                    result.error(
                        "invalid_call_id",
                        "No active incoming call found for callId=$callId",
                        null,
                    )
                    return
                }
                result.success(null)
            }

            "confirmAcceptedCall" -> {
                val callId = call.argument<String>(CallwaveConstants.EXTRA_CALL_ID)
                if (callId.isNullOrBlank()) {
                    result.error("invalid_call_id", "callId is required", null)
                    return
                }
                if (!callManager.confirmAcceptedCall(callId)) {
                    result.error(
                        "invalid_call_id",
                        "No active accepted call found for callId=$callId",
                        null,
                    )
                    return
                }
                result.success(null)
            }

            "declineCall" -> {
                val callId = call.argument<String>(CallwaveConstants.EXTRA_CALL_ID)
                if (callId.isNullOrBlank()) {
                    result.error("invalid_call_id", "callId is required", null)
                    return
                }
                if (!callManager.declineCall(callId)) {
                    result.error(
                        "invalid_call_id",
                        "No active incoming call found for callId=$callId",
                        null,
                    )
                    return
                }
                result.success(null)
            }

            "markMissed" -> {
                val callId = call.argument<String>(CallwaveConstants.EXTRA_CALL_ID)
                if (callId.isNullOrBlank()) {
                    result.error("invalid_call_id", "callId is required", null)
                    return
                }
                val extra = (call.argument<Map<*, *>>(CallwaveConstants.EXTRA_EXTRA))
                    ?.entries
                    ?.associate { it.key.toString() to it.value }
                callManager.markMissed(callId, extra)
                result.success(null)
            }

            "getActiveCallIds" -> {
                result.success(callManager.getActiveCallIds())
            }

            "getActiveCallEventSnapshots" -> {
                result.success(callManager.getActiveCallEventSnapshots())
            }

            "syncActiveCallsToEvents" -> {
                callManager.syncActiveCallsToEvents()
                result.success(null)
            }

            "takePendingStartupAction" -> {
                result.success(callManager.takePendingStartupAction())
            }

            "requestNotificationPermission" -> {
                result.success(callManager.requestNotificationPermission(activity))
            }

            "requestFullScreenIntentPermission" -> {
                callManager.requestFullScreenIntentPermission(activity)
                result.success(null)
            }

            "setPostCallBehavior" -> {
                val behavior = call.argument<String>(CallwaveConstants.EXTRA_POST_CALL_BEHAVIOR)
                callManager.setPostCallBehavior(behavior)
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun payloadFromCall(call: MethodCall): CallPayload? {
        val raw = call.arguments as? Map<*, *> ?: return null
        val normalized = raw.entries.associate { it.key.toString() to it.value }
        return try {
            CallPayload.fromMethodArgs(normalized)
        } catch (_: Throwable) {
            null
        }
    }
}
